apis/apiextensions/v1/composition_common.go

```diff
@@ -298,6 +298,9 @@ type PipelineStep struct {
 	// +kubebuilder:pruning:PreserveUnknownFields
 	// +kubebuilder:validation:EmbeddedResource
 	Input *runtime.RawExtension `json:"input,omitempty"`
+
+	// Credentials are optional credentials that the Composition Function needs.
+	Credentials *FunctionCredentials `json:"credentials,omitempty"`
```

Comment:
would it make sense to make this a list, I was kind of expecting that given that it's called `credentials` rather than `credential`

Reply:
Maybe. Can you think of a use-case for multiple sets of credentials? (e.g. Multiple secrets?)

I have a slight preference for keeping it a single set of credentials. That would match the ProviderConfig `credentials` API, which AFAIK has been good enough. I'm not aware of any asks to support multiple credential sets in that API.

P.S. A "credential" singular is pretty uncommon. From https://www.vocabulary.com/dictionary/credential:

> You’ll most often see the word credential used in its plural form: credentials. You typically refer to your credentials to show that you're qualified to do something.

For example I think a single username and password would be "credentials", not "a credential".

Reply:
True, worst case one can always put everything in a single secret at different keys. But should someone come up with a valid reason to get multiple secrets in the future, we would be stuck with a single one. I feel it makes sense for providers to only have a single set of credentials, but functions feel more flexible, hence the need could arise. But just thinking out loud, not really a strong opinion.

Reply:
How would we pass multiple credentials to the function? Would it make it more complicated for function authors (if we assume the common case will be either zero or one set of credentials).

Reply:
A map name:secret as we do for observed resources and extra resources?

Reply:
I casually just stumbled upon an example of a Provider that included a list of credentials in its ProviderConfig: https://github.com/crossplane-contrib/provider-ansible/blob/f07941702894f1209e7cb2763a5bdd1630158f3d/examples/provider/config.yaml#L15C1-L18C32

Reply:
I can see it being possible (maybe even likely) that there will be some functions that need to look up data from more than 1 external system and therefore need more than 1 set of credentials. I have a slight preference of supporting that in this API.

I don't **love** what that does for the common path of a function needing just 1 credentials, but instinct is that it will be needed by some functions - e.g. a function that looks up data in both GCP and AWS.

Reply:
Okay, I'm sold. 🙂 I've pushed a commit to support multiple credentials per function.